stats.py

```diff
@@ -1,17 +1,71 @@
 """Some simple statistical functions."""
 
-def mean(data):
-    """Takes the mean/average of a list of numbers, a simple measure of 'middle'. See: https://en.wikipedia.org/wiki/Arithmetic_mean"""
-    pass
+import numbers
+import collections
+import math
 
-def median(numbers):
-    """Finds the median or middle of a list of sorted numbers, a simple measure of 'middle'. See: https://en.wikipedia.org/wiki/Median"""
-    pass
+def mean(number_list):
+    """Takes the mean/average of a list of numbers, a simple measure of 'middle'.
+    See: https://en.wikipedia.org/wiki/Arithmetic_mean"""
+    #preconditions
+    assert isinstance(number_list, collections.Iterable), "Must be iterable."
+    for num in number_list:
+        assert isinstance(num, numbers.Number), "We need numbers yo."
+    assert len(number_list) > 1, "Please enter at least 2 numbers to be averaged."
+    #function
+    answer = (sum(number_list) / len(number_list))
+    #postconditions
+    assert isinstance(answer, numbers.Number), "Mean will return a number."
+    assert answer > min(number_list), "Mean should be higher than lowest number in list."
+    assert answer < max(number_list), "Mean should be lower than highest number in list."
+    return answer
 
-def range(numbers):
-    """Finds ths range of a list of numbers, a simple measure of variance. See: https://www.mathsisfun.com/definitions/range-statistics-.html"""
-    pass
+def median(number_list):
+    """Finds the median or middle of a list of sorted numbers, a simple measure of 'middle'.
+    See: https://en.wikipedia.org/wiki/Median"""
+    #preconditions
+    assert isinstance(number_list, collections.Iterable), "Must be iterable."
+    for num in number_list:
+        assert isinstance(num, numbers.Number), "We need numbers yo."
```

Comment:
Python list comprehensions can sometimes turn for loops into one liners
http://www.secnetix.de/olli/Python/list_comprehensions.hawk

they work especially well in preconditions/postconditions by combining them with any, all
https://docs.python.org/2/library/functions.html#all
https://docs.python.org/2/library/functions.html#any

Can you rewrite this and other for loop assertions using the above tools? something like

```
assert all(foo(x, ...) for x in number_list)
```

when complete, do you think it improves readability or not? why or why not?